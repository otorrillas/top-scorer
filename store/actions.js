import 'isomorphic-unfetch';

import _find from 'lodash/find';

import endpoints from '../services/endpoints.json';

export const actionTypes = {
  RESET: 'RESET',
  RESET: 'FAILURE',
  FETCH_GAMES_LIST: 'FETCH_GAMES_LIST',
  FETCH_GAMES_LIST_SUCCESS: 'FETCH_GAMES_LIST_SUCCESS',
  FETCH_RUNS: 'FETCH_RUNS',
  FETCH_RUNS_SUCCESS: 'FETCH_RUNS_SUCCESS'
};

export function reset() {
  return { type: actionTypes.RESET };
}

export function failure() {
  return { type: actionTypes.FAILURE };
}

export function fetchGamesList() {
  return async function(dispatch) {
    await fetch(endpoints.gameList)
      .then(response => response.json())
      .then(json => {
        dispatch(fetchGamesListSuccess(json.data));
      })
      .catch(error => console.log(error));
  };
}

export function fetchGamesListSuccess(data) {
  return {
    type: actionTypes.FETCH_GAMES_LIST_SUCCESS,
    data
  };
}

export function fetchRuns(id) {
  return async function(dispatch, getState) {
    const games = getState().games;

    const targetGame = _find(games, { id });
    const runsUrl = _find(targetGame.links, { rel: 'runs' }).uri;

    await fetch(runsUrl)
      .then(response => response.json())
      .then(json => {
        dispatch(fetchRunsSuccess(id, json.data));
      })
      .catch(error => console.log(error));
  };
}

export function fetchRunsSuccess(id, data) {
  return {
    type: actionTypes.FETCH_RUNS_SUCCESS,
    id,
    data
  };
}
