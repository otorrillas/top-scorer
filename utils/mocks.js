export const mockRuns = [
  {
    videos: { links: [{ uri: 'https://video-url' }] },
    players: [{ id: 'playerId' }],
    times: { primary_t: 450 }
  }
];

export const mockGames = {
  '1234': {
    abbreviation: 'test',
    assets: {
      'cover-medium': {
        uri: 'https://cover-medium-url'
      }
    },
    runs: mockRuns
  }
};

export const mockGamesResponse = [
  {
    id: '1234',
    ...mockGames['1234']
  },
  {
    id: '45678',
    ...mockGames['1234']
  }
];
