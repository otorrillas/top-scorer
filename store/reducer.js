import _keyBy from 'lodash/keyBy';

import { actionTypes } from './actions';

export const initialState = {
  error: null,
  games: [],
  selectedGame: null
};

function reducer(state = initialState, action) {
  switch (action.type) {
    case actionTypes.FAILURE:
      return {
        ...state,
        ...{ error: action.error }
      };

    case actionTypes.RESET:
      return {
        ...initialState
      };

    case actionTypes.FETCH_GAMES_LIST_SUCCESS:
      return {
        ...state,
        games: _keyBy(action.data, 'id')
      };

    case actionTypes.FETCH_RUNS_SUCCESS:
      return {
        ...state,
        games: {
          ...state.games,
          [action.id]: {
            ...state.games[action.id],
            runs: action.data
          }
        }
      };
    default:
      return state;
  }
}

export default reducer;
