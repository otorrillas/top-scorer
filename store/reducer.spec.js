import reducer from './reducer';
import { actionTypes } from './actions';

import { mockRuns, mockGamesResponse } from 'Utils/mocks';

describe('Redux - reducer', () => {
  test('failure sets error field', () => {
    expect(
      reducer([], {
        type: actionTypes.FAILURE,
        error: 'Some error'
      })
    ).toMatchSnapshot();
  });

  test('reset sets state to initial state', () => {
    expect(
      reducer(
        [
          {
            type: actionTypes.FAILURE,
            error: 'Some error'
          }
        ],
        {
          type: actionTypes.RESET
        }
      )
    ).toMatchSnapshot();
  });

  test('fetch game list success populates games', () => {
    expect(
      reducer([], {
        type: actionTypes.FETCH_GAMES_LIST_SUCCESS,
        data: mockGamesResponse
      })
    ).toMatchSnapshot();
  });

  test('fetch runs appends runs to games', () => {
    const id = mockGamesResponse[0].id;
    const result = reducer(
      [],
      { type: actionTypes.FETCH_GAMES_LIST_SUCCESS, data: mockGamesResponse },
      {
        type: actionTypes.FETCH_RUNS_SUCCESS,
        id,
        data: mockRuns
      }
    );

    expect(result).toMatchSnapshot();
    expect(result.games[id].runs).toBe(mockRuns);
  });
});
