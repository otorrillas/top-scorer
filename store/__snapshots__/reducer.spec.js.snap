// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux - reducer failure sets error field 1`] = `
Object {
  "error": "Some error",
}
`;

exports[`Redux - reducer fetch game list success populates games 1`] = `
Object {
  "games": Object {
    "1234": Object {
      "abbreviation": "test",
      "assets": Object {
        "cover-medium": Object {
          "uri": "https://cover-medium-url",
        },
      },
      "id": "1234",
      "runs": Array [
        Object {
          "players": Array [
            Object {
              "id": "playerId",
            },
          ],
          "times": Object {
            "primary_t": 450,
          },
          "videos": Object {
            "links": Array [
              Object {
                "uri": "https://video-url",
              },
            ],
          },
        },
      ],
    },
    "45678": Object {
      "abbreviation": "test",
      "assets": Object {
        "cover-medium": Object {
          "uri": "https://cover-medium-url",
        },
      },
      "id": "45678",
      "runs": Array [
        Object {
          "players": Array [
            Object {
              "id": "playerId",
            },
          ],
          "times": Object {
            "primary_t": 450,
          },
          "videos": Object {
            "links": Array [
              Object {
                "uri": "https://video-url",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`Redux - reducer fetch runs appends runs to games 1`] = `
Object {
  "games": Object {
    "1234": Object {
      "abbreviation": "test",
      "assets": Object {
        "cover-medium": Object {
          "uri": "https://cover-medium-url",
        },
      },
      "id": "1234",
      "runs": Array [
        Object {
          "players": Array [
            Object {
              "id": "playerId",
            },
          ],
          "times": Object {
            "primary_t": 450,
          },
          "videos": Object {
            "links": Array [
              Object {
                "uri": "https://video-url",
              },
            ],
          },
        },
      ],
    },
    "45678": Object {
      "abbreviation": "test",
      "assets": Object {
        "cover-medium": Object {
          "uri": "https://cover-medium-url",
        },
      },
      "id": "45678",
      "runs": Array [
        Object {
          "players": Array [
            Object {
              "id": "playerId",
            },
          ],
          "times": Object {
            "primary_t": 450,
          },
          "videos": Object {
            "links": Array [
              Object {
                "uri": "https://video-url",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`Redux - reducer reset sets state to initial state 1`] = `
Object {
  "error": null,
  "games": Array [],
  "selectedGame": null,
}
`;
